File: notes/realsh.py

import sympy as sp
from sympy import Symbol, re, im, I

# x = Symbol('x', real=True)
# y = Symbol('y', real=True)

# a basis function is defined by an l and an m tuple:
ll = [1, 3, 3, 4]
mm = [1, -2, -1, 1]

# the following function takes these tuples and converts them from
# a complex SH basis to a real SH basis
# def c2rsh(ll, mm):

# first we need the required symbols for

C = [ Symbol(f'C{j}', real=True) for j in range(0, len(ll)+1) ]
S = [ Symbol(f'S{j}', real=True) for j in range(0, len(ll)+1) ]
A = [ None for j in range(0, 2*len(ll)+1) ]
A[len(ll)] = C[0]
prodA = C[0]
for j in range(1, len(ll)+1):
    A[len(ll)-j] =  (1/sp.sqrt(2)) * (C[j] - I * S[j])
    A[len(ll)+j] =  ((-1)**j/sp.sqrt(2)) * (C[j] + I * S[j])
    prodA = prodA * A[len(ll)-j] * A[len(ll)+j]

c = Symbol('a', real=True) + I *  Symbol('b', real=True)
print(sp.simplify(sp.re(c * prodA)))

File: notes/realsh.jl

# --------------------------------------------------------------------------
# ACE.jl and SHIPs.jl: Julia implementation of the Atomic Cluster Expansion
# Copyright (c) 2019 Christoph Ortner <christophortner0@gmail.com>
# All rights reserved.
# --------------------------------------------------------------------------


using SymPy, StaticArrays

# a basis function is defined by an l and an m tuple:
ll = [1, 3, 3, 4]
mm = [1, -2, -1, 1]
c = rand() + im * rand()

function convert_AA(ll, mm, c)
    # length of the tuple / length of the product
    n = length(ll)

    # create symbols representing the real spherical harmonics
    # by analogy with exp(ikx) = cos(kx) + i sin(kx)
    C = [ symbols("C$j", real=true) for j in 1:length(ll) ]
    S = [ symbols("S$j", real=true) for j in 1:length(ll) ]

    # now express the complex spherical harmonics in terms of the real ones
    # (note we don't need to refer to the YR list since it stores the same
    #  symbols as the C, S lists)
    YR = Any[nothing for _=1:n]
    for j in 1:n
        if mm[j] == 0
            YR[j] = C[j]
        elseif mm[j] > 0
            YR[j] = (1/SymPy.sqrt(2)) * (C[j] - im * S[j])
        else
            YR[j] = ((-1)^mm[j]/SymPy.sqrt(2)) * (C[j] + im * S[j])
        end
    end

    # finally create the symbol for the prefactor
    a = symbols("a", real=true)
    b = symbols("b", real=true)
    coeff = a + im * b

    # let Sympy multiply and simplify the expression
    expr = simplify(real(coeff * prod(YR)))
    println(expr)

    # next, we need to extract the prefactors
    CS = [S; C]
    for ii in CartesianIndices(ntuple(_ -> length(CS), n))
        if length(unique(ii.I)) != length(ii); continue; end
        if sort([ii.I...]) != [ii.I...]; continue; end
        term = prod(CS[SVector(ii.I...)])
        pref = (expr.coeff(term)).subs(a, real(c)).subs(b, imag(c))
        if pref == 0; continue; end
        println(term, " -> ", pref)
    end
    nothing
end


convert_AA(ll, mm, c)
